### PR TITLE
with torch.no_grad

### PR DESCRIPTION
**Pull request by mavenlin**
When using `torch.no_grad`, an error is raised complaining about `torch._C.set_grad_enabled` is not implemented for `Torchish`.

This PR added a global variable to track whether gradient is enabled. Any new `Torchish` element created under the `no_grad` context will have its `.value` wrapped with `lax.stop_gradient`.

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -41,6 +41,9 @@ def RngPooperContext(value: RngPooper | None):
     _RNG_POOPER_STACK.pop()
 
 
+_GRAD_ENABLED = True
```
Comment (samuela) [inline torch2jax/__init__.py:44]: let's add a comment here documenting why this exists and what parts of the api it interacts with

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -84,6 +87,9 @@ def __init__(self, value):
     assert isinstance(value, jnp.ndarray) or isinstance(value, int) or isinstance(value, float), (
       f"Tried to create Torchish with unsupported type: {type(value)}"
     )
+    global _GRAD_ENABLED
+    if not _GRAD_ENABLED:
+      value = jax.lax.stop_gradient(value)
     self.value = value
```
Comment (samuela) [inline torch2jax/__init__.py:93]:
```suggestion
    self.value = value if _GRAD_ENABLED else jax.lax.stop_gradient(value)
```

**Review by samuela — COMMENTED** [review]
i'm hesitant to add a global variable but this may indeed be the simplest route. at some point we ought to also consider supporting/testing https://docs.pytorch.org/docs/stable/generated/torch.enable_grad.html, https://docs.pytorch.org/docs/stable/generated/torch.autograd.grad_mode.set_grad_enabled.html, and https://docs.pytorch.org/docs/stable/generated/torch.is_grad_enabled.html

**Comment by mavenlin** [comment]
BTW, I've changed `value` to be a property and added a `setter`, in this way, the inplace operations will work.

File: tests/test_core.py [hunk 3]
```diff
@@ -228,3 +228,65 @@ def f(x):
...
+      with torch.set_grad_enabled(False):
+        z = torch.cos(x)
+      z = z * 3
+    return y + z
+
+  t2j_function_test(f3, [()], atol=1e-6)
+
+  # test inplace functions
+  def f4(x):
+    # this is an effectively an identity function
```
Comment (samuela) [inline tests/test_core.py:286]:
```suggestion
    # this is effectively an identity function
```
nit

File: torch2jax/__init__.py [hunk 4]
```diff
@@ -507,6 +515,11 @@ def randperm(
   return jax.random.permutation(mk_rng(), n).astype(dtype or torch.int64)
 
 
+@implements(torch._C._set_grad_enabled, Torchishify_output=False)
+def set_grad_enabled(mode):
```
Comment (samuela) [inline torch2jax/__init__.py:519]:
```suggestion
def _set_grad_enabled(mode):
```
nit

**Review by samuela — COMMENTED** [review]
nice! reusing torch's internal thing is still global state but hey at least it's not our global state :)

**Comment by samuela** [comment]
ci should be fixed after b9588259ec7a2f63eed67d0c434086894a74c3e6

**Review by samuela — APPROVED** [review]
it's worth mentioning that
```py
with torch.no_grad():
  jax.grad(t2j(torch.sin))(x)
```
will be `None` or zero. it's debatable what should really happen here. i'm fine with merging as is and we can revisit later if necessary
